test_util/aws.py

```diff
@@ -274,46 +283,23 @@ def delete(self, delete_vpc=False):
             self.wait_for_stack_deletion()
             self.boto_wrapper.resource('ec2').Vpc(vpc_id).delete()
 
-    @retry_boto_rate_limits
-    def get_vpc(self):
-        vpc_id = self.boto_wrapper.client('ec2').describe_vpcs(Filters=[
-            {'Name': 'tag:Name', 'Values': [self.stack.stack_name]}])['Vpcs'][0]['VpcId']
-        return self.boto_wrapper.resouce('ec2').Vpc(vpc_id)
-
     def get_master_ips(self):
-        return self.get_substack_hosts('MasterStack')
+        return self.get_substack_ips('MasterStack')
 
     def get_private_agent_ips(self):
-        return self.get_substack_hosts('PrivateAgentStack')
+        return self.get_substack_ips('PrivateAgentStack')
 
     def get_public_agent_ips(self):
-        return self.get_substack_hosts('PublicAgentStack')
+        return self.get_substack_ips('PublicAgentStack')
 
-    @retry_boto_rate_limits
     def get_resource_stack(self, resource_name):
         """Returns a CfStack for a given resource
         """
-        resources = self.boto_wrapper.client('cloudformation').describe_stack_resources(StackName=self.stack.stack_name)
-        print(resources)
-        for r in resources['StackResources']:
-            if r['LogicalResourceId'] == resource_name:
-                return CfStack(r['PhysicalResourceId'], self.boto_wrapper)
-
-    @retry_boto_rate_limits
-    def get_substack_instance_ids(self, substack):
-        """returns list of instance ids
-        """
-        asg = self.get_resource_stack(substack).stack.\
-            Resource(substack.replace('Stack', 'ServerGroup')).physical_resource_id
-        return [i['InstanceId'] for i in self.boto_wrapper.client('autoscaling').
-                describe_auto_scaling_groups(
-                AutoScalingGroupNames=[asg])['AutoScalingGroups'][0]['Instances']]
+        return CfStack(self.stack.Resource(resource_name).physical_resource_id, self.boto_wrapper)
 
-    @retry_boto_rate_limits
-    def get_substack_hosts(self, substack):
-        instances = self.boto_wrapper.client('ec2').describe_instances(
-            InstanceIds=self.get_substack_instance_ids(substack))['Reservations'][0]['Instances']
-        return instances_to_hosts(instances)
+    def get_substack_ips(self, substack_name):
+        return instances_to_hosts(
+            self.get_resource_stack(substack_name).get_auto_scaling_instances(None))
```

Comment:
why can't we know the stack name here?

I really don't like the "sometimes we'll aggregate" for you behavior, in the instances where we're launching our templates, should always know exactly the AWS ASG resource name provided by the templates.